### DISCUSSION TITLE
leaked tracer error when calling neural net

### DISCUSSION THREAD

**Discussion by dnajera27** [google/flax#1635] [Q&A] [answered]
Hi, I keep getting a "leaked tracer" error when trying to run my model. I have read the documentation about what could be causing this error but I don't see how my code is triggering it. My application is pretty weird because I am trying to impose a Hamiltonian structure to my code, so I compute a bunch of jacobians and the Hessian of my neural network with respect to the inputs at various points. However, I am not sure that's the problem. I think I may be making a simple mistake in the way I wrote my code since I'm brand new to JAX. I'd appreciate any suggestions on how to solve this issue! 

[Colab notebook](https://colab.research.google.com/drive/1nA8ge8WW0U9_h2_jukAI2T7nPZ0m5Ryv?usp=sharing)

Thank you!

**Comment by jheek** [comment, marked as answer]
The JAX transformations like jacfwd/jacbwd, vmap, jit etc. don't work directly inside Modules. This is because they expect pure (stateless) functions. When you want to use JAX transforms inside Module you should use the stateful variants that flax provides e.g.: flax.nn.vmap, etc. I recently added stateful jacobian products which you can also use to compute the hessian. However, they aren't merged yet (https://github.com/google/flax/pull/1629)

**Reply by jheek** [reply]
More generally though, I think your code could be simplified by seperating the code that takes gradients/hessians and the paramaterized functions like the kinetic functions. Right now it's all mixed into one big Module. It's probably easier to make a Kinetic Module, potential Module, etc. And explicitly call them with `.apply(params, ...)`. This way you don't need the more complicated stateful transformations and you can use jax.hessian and friends out of the box.

**Reply by dnajera27** [reply]
Thanks for the reply!
So maybe something like this example?
https://gist.github.com/makora9143/0de98bcddd89a6a382222d16b70db9c5

Have a separate Kinetic and Potential class that inherit from nn.Module, and define all the functions that call jax.hessian, etc. separately instead of a single class?

Thanks!

**Reply by dnajera27** [reply]
And another question... can I still use TrainState.create() if I have three separate nn.Module classes? Right now I end up with three separate modules, so not sure what to pass to TrainState. Thanks!

**Reply by marcvanzee** [reply]
Do you use a single optimizer? Then it seems more appropriate to add all three Modules as submodules into a single parent Module.

**Reply by dnajera27** [reply]
I ended up writing a "dummy" module that calls the other modules to instantiate them for use in the TrainState. Thanks!
